rubicon/objc/api.py

```diff
@@ -656,6 +656,17 @@ def __new__(cls, object_ptr, _name=None, _bases=None, _ns=None):
 
         return self
 
+    def retain(self):
+        """
+        Manually increment the reference count of the corresponding objc object
+
+        The objc object is sent a dealloc message when its reference count reaches 0.
+        Use this method to retain objects which we don't own.
+        """
+        self._needs_release = True
+        result = send_message(self, "retain", restype=objc_id, argtypes=[])
```

Comment:
This unconditionally calls `retain` on the Objective-C side every time `retain` is called from Python, but the boolean `_needs_release` attribute can only track one of those `retain` calls (or none at all, if the object came from `alloc`/etc. and so already has `_needs_release` set to `True`). This means that if you retain an object multiple times from Python, `__del__` will not release all of those references, so you instead need to manually call `release` for every `retain` call.

This could be solved by replacing `_needs_release` with an integer reference counter, so that multiple `retain` calls can be remembered and `__del__` can call `release` the correct number of times.

Though I'm wondering if it's necessary at all to have custom handling for when `retain` is called from Python. The custom handling in `release` and `autorelease` is needed so that code that calls these methods continues to work, even though we added an automatic `release` call in `__del__` for Python code that never does Objective-C reference management. However, existing code that calls `retain` already works correctly with just the changes to `release`/`autorelease`, because this PR doesn't add any automatic calls to `retain` that would need to be suppressed if the user code already calls `retain` manually.

Reply:
True, that's a fair point. Rather than manually tracking the reference count in Python I am also inclined to just drop the `release` method override.